psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1413 @@
+                }
+            }
+
+            R_iajb_rms[ij] = R_iajb[ij]->rms();
+        }
+
+        // evaluate convergence using current amplitudes and residuals
+        e_prev = e_curr;
+        e_curr = eval_amplitudes();
+        r_curr = *max_element(R_iajb_rms.begin(), R_iajb_rms.end());
+
+        r_converged = (fabs(r_curr) < options_.get_double("R_CONVERGENCE"));
+        e_converged = (fabs(e_curr - e_prev) < options_.get_double("E_CONVERGENCE"));
+
+// use residuals to get next amplitudes
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            int i, j;
+            std::tie(i, j) = ij_to_i_j[ij];
+            for (int a = 0; a < n_pno[ij]; ++a) {
+                for (int b = 0; b < n_pno[ij]; ++b) {
+                    T_iajb[ij]->add(a, b, -R_iajb[ij]->get(a, b) / ((e_pno[ij]->get(a) + e_pno[ij]->get(b)) -
+                                                                    (F_lmo->get(i, i) + F_lmo->get(j, j))));
+                }
+            }
+        }
+
+        // DIIS extrapolation
+        SharedMatrix T_iajb_flat = flatten_mats(T_iajb);
+        SharedMatrix R_iajb_flat = flatten_mats(R_iajb);
+        if (iteration == 0) {
+            diis->set_error_vector_size(1, DIISEntry::Matrix, R_iajb_flat.get());
+            diis->set_vector_size(1, DIISEntry::Matrix, T_iajb_flat.get());
+        }
+        diis->add_entry(R_iajb_flat, T_iajb_flat);
+        diis->extrapolate(T_iajb_flat);
+        copy_flat_mats(T_iajb_flat, T_iajb);
+
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            Tt_iajb[ij]->zero();
+            Tt_iajb[ij]->add(T_iajb[ij]);
+            Tt_iajb[ij]->scale(2.0);
+            Tt_iajb[ij]->subtract(T_iajb[ij]->transpose());
+        }
+
+        outfile->Printf("  @LMP2 iter %3d: %16.12f %10.3e %10.3e\n", iteration, e_curr, e_curr - e_prev, r_curr);
+
+        iteration++;
+    }
+
+    e_lmp2_ = e_curr;
+}
+
+double DLPNOMP2::eval_amplitudes() {
+    double e_mp2 = 0.0;
+    for (int ij = 0; ij < Tt_iajb.size(); ++ij) {
+        e_mp2 += K_iajb[ij]->vector_dot(Tt_iajb[ij]);
+        e_mp2 += R_iajb[ij]->vector_dot(Tt_iajb[ij]);
+    }
+    return e_mp2;
+}
+
+void DLPNOMP2::setup() {
+    int natom = molecule_->natom();
+    int nbf = basisset_->nbf();
+    int nshell = basisset_->nshell();
+    int naux = ribasis_->nbf();
+    int nshellri = ribasis_->nshell();
+    int naocc = nalpha_ - basisset_->n_frozen_core(options_.get_str("FREEZE_CORE"), molecule_);
+
+    SharedMatrix C_vir = reference_wavefunction_->Ca_subset("AO", "VIR");
+    SharedVector e_vir = reference_wavefunction_->epsilon_a_subset("AO", "VIR");
+
+    SharedMatrix C_occ = reference_wavefunction_->Ca_subset("AO", "OCC");
+    SharedVector e_occ = reference_wavefunction_->epsilon_a_subset("AO", "OCC");
+
+    // Localize active occupied orbitals
+    auto localizer = std::make_shared<BoysLocalizer>(
```

Comment:
Why do we insist on a Boys localizer as opposed to a PM localizer? I understand that the original paper used Boys, but ORCA supports both and lets the user choose, if they want.

Reply:
I added a `DLPNO_LOCAL_ORBITALS` option that defaults to `BOYS` but may also be `PIPEK_MEZEY`. The localization scheme shouldn't matter much according to Neese and coworkers (https://doi.org/10.1063/1.3086717) who say "The default localization method is the Foster–Boys procedure that we find to be more robust than the Pipek–Mezey method that is also available. However, the results of the calculations depend very little on the actual localization method used"